### PR TITLE
Add an option to allow lowercase hexadecimal.

### PR DESCRIPTION
**Pull request by almet**
Here is a way to lowercase the generated codepoints (useful for python compatibility for instance).

### CODE REVIEW

**Comment by mathiasbynens** [comment]
Why should this be added?

Python compatibility is not an issue since Python supports both `\xE9` and `\xe9`, to use your example.

**Comment by almet** [comment]
Thanks for your answer. Actually, python supports lowercase and uppercase hexadecimal as input, but not as output: it always outputs lowercase.

My use case is to be able to serialize JSON in a canonical way between python and JavaScript (at first), and for this I need to have the same representation of unicode chars (otherwise the canonical form differs).

One way is of course to change the output of python. I think both approach are valid, and thought that this use case could fit others needs, hence the contribution.

Is the intent clearer here?

File: jsesc.js [hunk 1]
```diff
@@ -225,7 +230,10 @@
 				continue;
 			}
 			var charCode = character.charCodeAt(0);
-			var hexadecimal = charCode.toString(16).toUpperCase();
+      var hexadecimal = charCode.toString(16);
```
Comment (mathiasbynens) [inline jsesc.js]: Please follow the existing coding style (tabs for indentation, etc.).

File: tests/tests.js [hunk 2]
```diff
@@ -287,6 +287,13 @@
 			'{"hello":"world","\\uD83D\\uDCA9":"foo","pile":"\\uD83D\\uDCA9"}',
 			'`toJSON` methods are not called when `json: false`'
 		);
+    equal(
+			jsesc('→é', {
```
Comment (mathiasbynens) [inline tests/tests.js]: No non-ASCII symbols in the source code. Please escape these.

File: tests/tests.js [hunk 3]
```diff
@@ -287,6 +287,13 @@
...
+				'toUpperCase': false
+			}),
+			'\\u2192\\xe9',
+			'Code points are lowercase when `toUpperCase: false`'
```
Comment (mathiasbynens) [inline tests/tests.js]: More accurately: “alphabetical hexadecimal digits are lowercase when `toUpperCase: false`”.

Comment (mathiasbynens) [inline tests/tests.js]: Please add some more tests combining this setting with the other options.

**Comment by mathiasbynens** [comment]
I left some inline comments. The documentation would need to be updated as well.

File: jsesc.js [hunk 4]
```diff
@@ -95,6 +95,7 @@
 			'es6': false,
 			'json': false,
 			'compact': true,
+			'toUpperCase': true,
```
Comment (mathiasbynens) [inline jsesc.js]: I’m not sure `toUpperCase` is the best option name — how about `upperCaseHex` instead?

**Comment by almet** [comment]
I tried to address your concerns, let me know if that's enough, I'm happy to do some more changes if needed.

**Comment by mathiasbynens** [comment]
Looking good so far!

Nit: let’s go for `uppercaseHex` instead of `upperCaseHex` since “uppercase” is one word.

Thought of some more things — sorry I forgot to mention these earlier:

- [ ] The CLI (see `bin` directory) should get an option to enable this setting. I’d propose `-u | --uppercase-hex`.
- [ ] The `man` page should be updated accordingly.

**Comment by almet** [comment]
Now that I think about it, maybe the option should be `lowercaseHex`, because by default it's upper case. What do you think?

**Comment by almet** [comment]
I've just updated the option to expose `lowercaseHex` rather than `uppercaseHex`.

File: README.md [hunk 5]
```diff
@@ -348,6 +360,13 @@ Or do the same with an online JSON file:
 $ curl -sL "http://git.io/aorKgQ" | jsesc --json --object > data-escaped.json
 ```
 
+For lowercase hexadecimal values, use the `-l`/`--lowercase-hex` option:
```
Comment (mathiasbynens) [inline README.md]: This section can be removed — only a few examples of common usage are listed here.

File: README.md [hunk 6]
```diff
@@ -348,6 +360,13 @@ Or do the same with an online JSON file:
...
+
+```bash
+$ jsesc --lowercase-hex 'föo ♥ bår 𝌆 baz'
+f\xf6o \u2665 b\xe5r \ud834\udf06 baz
```
Comment (mathiasbynens) [inline README.md]: Same here.

**Comment by mathiasbynens** [comment]
Good call on `lowercaseHex` vs `uppercaseHex`!

**Comment by almet** [comment]
Updated!

**Comment by almet** [comment]
Hey! Thanks for your time again. Anything else I can do to land this?

**Comment by mathiasbynens** [comment]
Please squash your commits.

I’m going over it one last time.

File: README.md [hunk 7]
```diff
@@ -290,6 +290,18 @@ jsesc([ undefined, -Infinity ], {
...
 
+#### `lowercaseHex`
+
+The `lowercaseHex` option takes a boolean value (`true` or `false`), and defaults to `false`. When enabled, the hexadecimal values in the output are in lower case.
+
+```js
+jsesc('Ich ♥ Bücher', {
+  'lowercaseHex': true,
+});
+// → 'Ich \u2665 B\xfccher'
```
Comment (mathiasbynens) [inline README.md]: needs double `\\`

File: tests/tests.js [hunk 8]
```diff
@@ -287,6 +287,60 @@
 			'{"hello":"world","\\uD83D\\uDCA9":"foo","pile":"\\uD83D\\uDCA9"}',
 			'`toJSON` methods are not called when `json: false`'
 		);
+		equal(
+			jsesc('\u2192\xe9', {
```
Comment (mathiasbynens) [inline tests/tests.js]: Nit: uppercase hex digits

Comment (mathiasbynens) [inline tests/tests.js]: same for the others below

**Comment by mathiasbynens** [comment]
LGTM with those minor nits addressed + commits squashed. :+1:

**Comment by almet** [comment]
Just updated and squashed!

File: tests/tests.js [hunk 9]
```diff
@@ -287,6 +287,60 @@
...
 			'`toJSON` methods are not called when `json: false`'
 		);
+		equal(
+			jsesc('\u2192\xE9', {
+				'lowercaseHex': true
+			}),
+			'\\u2192\\xe9',
+			'Alphabetical hexadecimal digits are lowercase when `lowercaseHex: true`'
+		);
+		equal(
```
Comment (mathiasbynens) [inline tests/tests.js]: I just noticed there’s only a single test with `lowercaseHex: true`. From this test on, please duplicate each of these tests with `lowercaseHex: true`.

**Comment by mathiasbynens** [comment]
Are you up for adding those tests? If not, I’ll merge this and do it myself later before releasing a new version.

**Comment by almet** [comment]
Oh, this is because of the change of setting! Is it okay to switch them to true for all the tests added (that was the original intent)

**Comment by almet** [comment]
I've updated the tests with the two cases "true" and "false". Let me know if this is enough :)

File: tests/tests.js [hunk 10]
```diff
@@ -287,6 +287,100 @@
...
+			'"\\u2192\\u00E9"',
+			'Alphabetical hexadecimal digits are uppercase when `lowercaseHex: alse` and `json: true`'
+		);
+		equal(
+			jsesc('test', {
+				'lowercaseHex': false,
+				'escapeEverything': true
+			}),
+			'\\x74\\x65\\x73\\x74',
+			'Alphabetical hexadecimal digits are uppercase when `lowercaseHex: false` and `escapeEverything: true`'
```
Comment (mathiasbynens) [inline tests/tests.js]: No alphabetical digits are being tested here

File: tests/tests.js [hunk 11]
```diff
@@ -287,6 +287,100 @@
...
+			}),
+			'"\\u2192\\u00e9"',
+			'Alphabetical hexadecimal digits are lowercase when `lowercaseHex: alse` and `json: true`'
+		);
+		equal(
+			jsesc('test', {
+				'lowercaseHex': true,
+				'escapeEverything': true
+			}),
+			'\\x74\\x65\\x73\\x74',
```
Comment (mathiasbynens) [inline tests/tests.js]: No alphabetical digits are being tested here

Comment (mathiasbynens) [inline tests/tests.js]: Maybe use a string like [`'I\xF1t\xEBrn\xE2ti\xF4n\xE0liz\xE6ti\xF8n\u2603\uD83D\uDCA9'`](https://mathiasbynens.be/notes/javascript-unicode#poo-test)?

**Comment by almet** [comment]
Updated.

File: tests/tests.js [hunk 12]
```diff
@@ -287,6 +287,100 @@
...
+		equal(
+			jsesc('\u2192\xe9', {
+				'lowercaseHex': true,
+				'json': true
+			}),
+			'"\\u2192\\u00e9"',
+			'Alphabetical hexadecimal digits are lowercase when `lowercaseHex: alse` and `json: true`'
+		);
+		equal(
+			jsesc('ççaçç', {
```
Comment (mathiasbynens) [inline tests/tests.js]: This should be escaped as `'\xE7\xE7a\xE7\xE7'` (no non-ASCII in source code)

**Comment by mathiasbynens** [comment]
Also, the `man` page entry still has `uppercase-hex`. I’ll merge and fix those nits myself.

**Comment by mathiasbynens** [comment]
Thanks!

**Comment by mathiasbynens** [comment]
v1.0.0 includes this change.

**Pull request closed without merging**